src/Processors/Formats/IRowInputFormat.h

```diff
@@ -29,6 +29,7 @@ struct RowInputFormatParams
 
     Poco::Timespan max_execution_time = 0;
     OverflowMode timeout_overflow_mode = OverflowMode::THROW;
+    bool sync_after_error = false;
```

Comment:
That looks unused. Why needed?

Reply:
It's needed. We exception occurs, the format object should skip  to next line or EOF before the exception throwed.
see the function syncAfterError().

Reply:
1) I don't see it used:
```
$ rg sync_after_error 
src/Formats/FormatFactory.cpp
273:    const bool sync_after_error) const
293:    params.sync_after_error = sync_after_error;

src/Formats/FormatFactory.h
137:        const bool sync_after_error = false) const;

src/Processors/Formats/IRowInputFormat.h
32:    bool sync_after_error = false;
```
2) my expectations: after exception we stop parsing the message as a whole, and do resetParser to clean the state to be able to start parsing new message (and we don't care too much if some CR / LF was skipped in the previous msg).

Reply:
at src/Storages/Kafka/KafkaBlockInputStream.cpp:87, initialize the sync_after_error to `put_error_to_stream.
Before reset the parser, we should call syncAfterError() to avoid the exception in the resetParser function.

> "Unread data in IInputFormat::resetParser. Most likely it's a bug."

Reply:
Hm... I'm wondering if we still need that assert there...

Reply:
I see now you added part src/Processors/Formats/IRowInputFormat.cpp :) It was missing before, and that was confusing me. 

Adding some a flag for src/Processors/Formats/IRowInputFormat.cpp will not help to solve the explained issue in general case (not all formats are based on IRowInputFormat).

Please try to create the following test: create a message in format=Native (having several rows & several columns).

```
clickhouse-local -q "select toUInt8(1) as a, toUInt16(2) as b, '\xDE\xAD\xBE\xEF' as c, toFixedString('\xCA\xFE\xBA\xBE',4) as d from numbers(5) format Native" | xxd 
00000000: 0405 0161 0555 496e 7438 0101 0101 0101  ...a.UInt8......
00000010: 6206 5549 6e74 3136 0200 0200 0200 0200  b.UInt16........
00000020: 0200 0163 0653 7472 696e 6704 dead beef  ...c.String.....
00000030: 04de adbe ef04 dead beef 04de adbe ef04  ................
00000040: dead beef 0164 0e46 6978 6564 5374 7269  .....d.FixedStri
00000050: 6e67 2834 29ca feba beca feba beca feba  ng(4)...........
00000060: beca feba beca feba be                   .........
```

and 'brake' it in few different ways 
1) change the type `String` to some unrecognized, like `Spring`
2) crop half of last column, 1 + half etc. 

Produce those messages and try to parse them.

Reply:
> Hm... I'm wondering if we still need that assert there...

If the assert removed, the flag `sync_after_error is not needed.

Reply:
Let's remove that assert or add a flag there like 'resetParser(after_exception)'

Reply:
@filimonov Hi, the assert is deleted.